rpc/RPCSession.java

```diff
@@ -214,37 +211,105 @@ public String database() {
+                            key -> {
+                                LOG.debug("Opening a session to '{}'", key);
+                                return clusterClient.coreClient(key.address()).session(key.database(), this.type, this.options);
+                            }
                     );
+                    LOG.debug("Opening read secondary transaction to secondary replica '{}'", replica);
+                    return selectedSession.transaction(type, options);
+                } catch (GraknClientException e) {
+                    if (e.getErrorMessage().equals(UNABLE_TO_CONNECT)) {
+                        LOG.debug("Unable to open a session or transaction to " + replica.id() + ". Reattempting to the next one.", e);
+                    } else {
+                        throw e;
+                    }
+                }
+            }
+            throw clusterNotAvailableException();
+        }
+
+        private Database databaseDiscover() {
+            for (Address.Cluster.Server server: clusterClient.servers()) {
+                try {
+                    return databaseDiscover(server);
                 } catch (StatusRuntimeException e) {
-                    GraknClusterGrpc.GraknClusterBlockingStub old = clusterDiscoveryRPC;
-                    clusterDiscoveryRPC = this.clusterClient.selectNextClusterDiscoveryRPC();
-                    LOG.info("Cluster Discovery RPC failed - server '{}' not available. Attempting to perform the operation onto another server '{}'",
-                            old.getChannel().authority(), clusterDiscoveryRPC.getChannel().authority());
+                    LOG.debug("Unable to perform database discovery to " + server + ". Reattempting to the next one.", e);
                 }
             }
+            throw clusterNotAvailableException();
+        }
+
+        private Database databaseDiscover(Address.Cluster.Server server) {
+            return Database.ofProto(
+                    clusterClient.graknClusterRPC(server).databaseDiscover(
+                        DatabaseProto.Database.Discover.Req.newBuilder()
+                                .setDatabase(dbName)
+                                .build()
+                    )
+            );
+        }
+
+        private GraknClientException clusterNotAvailableException() {
+            String addresses = clusterClient.servers().stream().map(Address.Cluster.Server::toString).collect(Collectors.joining(","));
+            return new GraknClientException(CLUSTER_UNABLE_TO_CONNECT, addresses); // remove ambiguity by casting to Object
+        }
 
-            throw new GraknClientException(CLUSTER_NOT_AVAILABLE.message(clusterClient.coreClients()));
+        private void sleepWait() {
+            try {
+                Thread.sleep(2000);
+            } catch (InterruptedException e2) {
+                throw new GraknClientException(UNEXPECTED_INTERRUPTION);
+            }
         }
 
         public static class Database {
```

Comment:
Is this class distinct from the concept of a _database_ (i.e. a knowledge graph?) If so, do you think we can disambiguate?

Reply:
Yes, it is distinct from the concept of a "Grakn database" but I think it should be fine since it's already contextualised to be within `RPCSession`.

The name is consistent with the name that we are using in the protocol definition: https://github.com/graknlabs/protocol/blob/master/protobuf/cluster/database.proto.

I do feel that something here can be improved, but it would have to be the overall structure rather than just this one part. I think we should do it as part of the major refactor we're going to do after we're done with Cluster alpha.

What do you think @alexjpwalker ?

Reply:
Perfectly fine with that